### PR TITLE
test-nss: allow the module and names/addresses to be specified

### PR DESCRIPTION
**Pull request by keszybz**
Useful for testing a single module. If nothing is specified,
behaviour is the same as before.

@martinpitt

### CODE REVIEW

File: src/test/test-nss.c [hunk 1]
```diff
@@ -450,5 +451,117 @@ int main(int argc, char **argv) {
                 log_info(" ");
         }
 
+        return 0;
+}
+
+static char** arg_modules = NULL;
+static char** arg_names = NULL;
```
Comment (poettering) [inline src/test/test-nss.c]: I'd always place static variables like this at the top of the .c file... But this doesn't really matter...

File: src/test/test-nss.c [hunk 2]
```diff
@@ -450,5 +451,117 @@ int main(int argc, char **argv) {
...
+                                if (!x)
+                                        return log_oom();
+
+                                *name = x;
+                        }
+                }
+        } else {
+                char *hostname;
+
+                names = strv_split_nulstr("localhost\0" "gateway\0" "foo_no_such_host\0");
```
Comment (poettering) [inline src/test/test-nss.c]: not that it mattrers, but there's also `strv_new("localhost", "gateway", "foo_no_such_host", NULL)` which might be a bit more readable.

File: src/test/test-nss.c [hunk 3]
```diff
@@ -450,5 +451,117 @@ int main(int argc, char **argv) {
...
+                if (!names)
+                        return log_oom();
+
+                hostname = gethostname_malloc();
+                if (!hostname)
+                        return log_oom();
+
+                r = strv_push(&names, hostname);
```
Comment (poettering) [inline src/test/test-nss.c]: strv_consume(), no? otherwise memleak...

File: src/test/test-nss.c [hunk 4]
```diff
@@ -450,5 +451,117 @@ int main(int argc, char **argv) {
...
+                if (r < 0)
+                        return r;
+
+                n = make_addresses(addresses);
+                if (n < 0)
+                        return n;
+        }
+
+        assert_se(strv_remove(names, "__ignore__"));
```
Comment (poettering) [inline src/test/test-nss.c]: not sure I grok the **ignore** logic?

**Comment by poettering** [comment]
Looks vaguely OK to me, but see comments. I figure @martinpitt should comment on this before we merge this?

File: src/test/test-nss.c (hunk 4 above)
Comment (keszybz) [reply src/test/test-nss.c]: I wanted to remove some items from strv while iterating over it. Is there a better way to do this?

Comment (poettering) [reply src/test/test-nss.c]: ah, got it now. But you are currently leaking the original entry, no?

For some loops like this where I wanted to drop entries in the middle while iterating through it, I opted for a plain for loop, with two index counters f, t, both starting at 0. For each entry that should stay I'd copy the entry at position f to position t, all in place, and would increase both indexes. But if I wanted to drop an entry I'd free the entry at position f, and only increase f, but leave t as it is.

This changes the array in place, and in the end has the filtered array only.

**Comment by poettering** [comment]
looks ok, but two memleaks, see above

**Comment by keszybz** [comment]
Rewritten based on comments.

**Pull request merged**